### PR TITLE
Use a separate GHA cache for each matrix build

### PR DESCRIPTION
**Pull request by richvdh**
You can't update a GHA cache once you create it, so if we use the same cache
for each job in a matrix build, then we end up populating it for the first job
that completes, so any slower jobs don't get their dependencies cached.

On the other hand, if we create 20 500MB cache items on each build, we're going
to exhaust the cache storage as soon as we do a build. So, instead, let's just
do the caching for the main branch, and hope that other branches can still
benefit from it.

~~Based on https://github.com/matrix-org/matrix-rust-sdk/pull/1607~~

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -47,6 +47,14 @@ jobs:
...
       - name: Load cache
         uses: Swatinem/rust-cache@v2
+        with:
+          # use a separate cache for each job to work around
+          # https://github.com/Swatinem/rust-cache/issues/124
+          key: "${{ matrix.name }}"
+
+          # ... but only save the cache on the main branch
+          # cf https://github.com/Swatinem/rust-cache/issues/95
+          save-if: ${{ github.ref == 'refs/head/master' }}
```
Comment (jplatte) [inline .github/workflows/ci.yml:57]: Our "master" branch is named `main`.

Comment (richvdh) [reply .github/workflows/ci.yml:57]: indeed. I am an idiot.

**Review by jplatte — APPROVED** [review]
Let's try it!

**Comment by richvdh** [comment]
overriding branch protection because codecov is broken...

**Pull request merged**